src/OmniSharp.Roslyn.CSharp/Services/Diagnostics/ReAnalyzeService.cs

```diff
@@ -0,0 +1,81 @@
+using System;
+using System.Collections.Immutable;
+using System.Composition;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.Extensions.Logging;
+using OmniSharp.Abstractions.Models.V1.ReAnalyze;
+using OmniSharp.Mef;
+using OmniSharp.Models;
+using OmniSharp.Models.Diagnostics;
+using OmniSharp.Roslyn.CSharp.Workers.Diagnostics;
+
+namespace OmniSharp.Roslyn.CSharp.Services.Diagnostics
+{
+    [OmniSharpHandler(OmniSharpEndpoints.ReAnalyze, LanguageNames.CSharp)]
+    public class ReAnalyzeService : IRequestHandler<ReAnalyzeRequest, ReanalyzeResponse>
+    {
+        private readonly ICsDiagnosticWorker _diagWorker;
+        private readonly OmniSharpWorkspace _workspace;
+        private readonly ILogger<ReAnalyzeService> _logger;
+
+        [ImportingConstructor]
+        public ReAnalyzeService(ICsDiagnosticWorker diagWorker, OmniSharpWorkspace workspace, ILoggerFactory loggerFactory)
+        {
+            _diagWorker = diagWorker;
+            _workspace = workspace;
+            _logger = loggerFactory.CreateLogger<ReAnalyzeService>();
+        }
+
+        public Task<ReanalyzeResponse> Handle(ReAnalyzeRequest request)
+        {
+
+            if(!string.IsNullOrEmpty(request.CurrentOpenFilePathAsContext))
+            {
+                var currentSolution = _workspace.CurrentSolution;
+
+                var projectIds = WhenRequestIsProjectFileItselfGetFilesFromIt(request.CurrentOpenFilePathAsContext, currentSolution)
+                    ?? GetDocumentsFromProjectWhereDocumentIs(request.CurrentOpenFilePathAsContext, currentSolution);
+
+                _logger.LogInformation($"Queue analysis for project(s) {string.Join(", ", projectIds)}");
+
+                _diagWorker.QueueDocumentsOfProjectsForDiagnostics(projectIds);
+            }
+            else
+            {
+                _logger.LogInformation($"Queue analysis for all projects.");
+                _diagWorker.QueueAllDocumentsForDiagnostics();
+            }
+
+            return Task.FromResult(new ReanalyzeResponse());
+        }
+
+        private ImmutableArray<ProjectId>? WhenRequestIsProjectFileItselfGetFilesFromIt(string currentOpenFilePathAsContext, Solution currentSolution)
+        {
+            var projects = currentSolution.Projects.Where(x => CompareProjectPath(currentOpenFilePathAsContext, x)).Select(x => x.Id).ToImmutableArray();
+
+            if(!projects.Any())
+                return null;
+
+            return projects;
+        }
+
+        private static bool CompareProjectPath(string currentOpenFilePathAsContext, Project x)
+        {
+            return String.Compare(
+                x.FilePath,
+                currentOpenFilePathAsContext,
+                StringComparison.InvariantCultureIgnoreCase) == 0;
+        }
+
+        private static ImmutableArray<ProjectId> GetDocumentsFromProjectWhereDocumentIs(string currentOpenFilePathAsContext, Solution currentSolution)
```

Comment:
this method name seems weird. it says `GetDocumentsFromProjectWhereDocumentIs` whereas in reality it is something more of a `GetProjectIdsFromDocumentFilePaths`